### DISCUSSION TITLE
CM4 U-Boot Patches

### DISCUSSION THREAD

**Discussion by lovesegfault** [home-assistant/operating-system#1871] [General]
Hi @agners!

While working on https://github.com/NixOS/nixpkgs/pull/170528, adding support for booting NixOS on the CM4 using U-Boot, I came across the incredibly useful patches here. Without them, I was unable to get U-Boot to work on the CM4 when booting off USB or NVMe.

I have a couple of questions regarding the future of the patches:

1. Are you planning to upstream this work?
2. Have you tried rebasing the patches for U-Boot 2022.04? I tried this, but was unable to make it work. After rebasing the patches, U-Boot can see the NVMe device on the PCIe interface, but it does not recognize it as an NVMe device, so booting fails.

Thank you!

**Comment by agners** [comment]
> 1. Are you planning to upstream this work?

I did send [an RFC](https://www.mail-archive.com/u-boot@lists.denx.de/msg418262.html) of that patch set a while ago. There is one hack which I am not sure about, but I did not receive feedback, so maybe its fine? I'll send an rebased PATCH version once I come around.

> 2. Have you tried rebasing the patches for U-Boot 2022.04? I tried this, but was unable to make it work. After rebasing the patches, U-Boot can see the NVMe device on the PCIe interface, but it does not recognize it as an NVMe device, so booting fails.

I just yesterday rebased them on 2022.01 with #1868. I haven't rebased to newer U-Boot versions yet.

From your description it sounds like translation issues again. Maybe that hack described above bites back :cold_sweat:

**Reply by lovesegfault** [reply]
> I did send [an RFC](https://www.mail-archive.com/u-boot@lists.denx.de/msg418262.html) of that patch set a while ago. There is one hack which I am not sure about, but I did not receive feedback, so maybe its fine? I'll send an rebased PATCH version once I come around.

I asked around IRC, and it seems like that patchset was just missed, so perhaps re-submitting will be sufficient.

I'm happy to add a Tested-By to confirm it solves a real issue.

> I just yesterday rebased them on 2022.01 with https://github.com/home-assistant/operating-system/pull/1868. I haven't rebased to newer U-Boot versions yet.

> From your description it sounds like translation issues again. Maybe that hack described above bites back :cold_sweat:

I think some of the changes you made, namely `nvme: translate virtual addresses into the bus's address`, touched parts of the code that have changed meaningfully.

For example, in your patch you do:
```diff
diff --git a/drivers/nvme/nvme.c b/drivers/nvme/nvme.c
index 6ab94ada7e..dd190bd654 100644
--- a/drivers/nvme/nvme.c
+++ b/drivers/nvme/nvme.c
@@ -835,6 +838,7 @@ static int nvme_probe(struct udevice *udev)
 	struct nvme_id_ns *id;
 
 	ndev->instance = trailing_strtol(udev->name);
+	ndev->dev = udev->parent;
 
 	INIT_LIST_HEAD(&ndev->namespaces);
 	ndev->bar = dm_pci_map_bar(udev, PCI_BASE_ADDRESS_0,

```

But upstream, this chunk looks like:
```c
	struct nvme_dev *ndev = dev_get_priv(udev);
	struct nvme_id_ns *id;
	int ret;

	ndev->udev = udev;
	INIT_LIST_HEAD(&ndev->namespaces);
	if (readl(&ndev->bar->csts) == -1) {
		ret = -ENODEV;
		printf("Error: %s: Out of memory!\n", udev->name);
		goto free_nvme;
	}

```

I imagine `nvme_virt_to_bus` will need to be modified accordingly.

Alas, despite playing with the patches, I've had no luck with `2022.04`.
